backend/accounting-service/src/main/kotlin/services/projects/v2/ProjectService.kt

```diff
@@ -1245,6 +1248,272 @@ class ProjectService(
+                        where project = :project and
+                            id in (select unnest(cast(:groups as text[])))
+                    ), changed as (
+                        insert into project.invite_link_group_assignments (group_id, link_token)
+                        select group_id, token from new
+                        on conflict do nothing
+                    ) delete from project.invite_link_group_assignments
+                    where link_token = cast(:token as uuid) and
+                        (group_id, link_token) not in (select group_id, token from new)
+                """
+            )
+        }
+    }
+
+    suspend fun acceptInviteLink(
+        actorAndProject: ActorAndProject,
+        request: ProjectsAcceptInviteLinkRequest,
+        ctx: DBContext = db
+    ) : ProjectsAcceptInviteLinkResponse {
+        data class ProjectInviteLinkAssignment(
+            val piActorAndProject: ActorAndProject,
+            val role: ProjectRole,
+            val groups: List<String>
+        )
+
+        return ctx.withSession { session ->
+            val projectAssignment = session.sendPreparedStatement(
+                {
+                    setParameter("token", request.token)
+                },
+                """
+                    select m.project_id as project, m.username as pi, l.role_assignment as role, a.group_id as group_id
+                    from project.project_members m
+                    inner join project.invite_links l on
+                        now() < expires and
+                        token = cast(:token as uuid) and
+                        l.project_id = m.project_id
+                    left join project.invite_link_group_assignments a on
+                        link_token = cast(:token as uuid)
+                    where
+                        m.role = 'PI';
+                """
+            ).rows.groupBy { it.getString("project") }.map { entry ->
+                val project = entry.key
+                val pi = entry.value.firstOrNull()?.getString("pi")
+                val role = entry.value.firstOrNull()?.getString("role")
+
+                if (project == null || pi == null || role == null) {
+                    throw RPCException("Link expired", HttpStatusCode.BadRequest)
+                }
+
+                ProjectInviteLinkAssignment(
+                    ActorAndProject(
+                        Actor.SystemOnBehalfOfUser(pi),
+                        project
+                    ),
+                    ProjectRole.valueOf(role),
+                    entry.value.mapNotNull { it.getString("group_id") }
+                )
+            }.firstOrNull() ?: throw RPCException("Link expired", HttpStatusCode.BadRequest)
+
+            createInvite(
+                projectAssignment.piActorAndProject,
+                bulkRequestOf(ProjectsCreateInviteRequestItem(actorAndProject.actor.safeUsername())),
+                ctx
+            )
+
+            acceptInvite(
```

Comment:
As far as I can tell, this has the unfortunate side-effect of sending an email to the user that they have been invited to the project. We should obviously not send such an email. I suggest we add a new parameter to `createInvite` and `changeRole` which disables notifications and emails.